### PR TITLE
Event versioning strategies

### PR DESCRIPTION
**Pull request by pienikrz**

### CODE REVIEW

File: src/main/java/com/hltech/store/versioning/MappingBasedVersioning.java [hunk 1]
```diff
@@ -3,7 +3,23 @@
...
+ * In this strategy, every event exists only in latest version, so that the application code has to support only one version of the event.
+ * The mapping strategy is based on three simple principles:
+ * - When attribute exists on both json and class then set the value from json
+ * - When attribute exists on json but not on class then do nothing
+ * - When attribute exists on class but not in json then set default value
+ *
+ * <p>This strategy is recommended when you have multiple instance of you application running at the same time,
+ * because it supports backward and forward compatibility.
+ *
+ * <p>Be aware, that it also has one important and annoying drawback. You are no longer allowed to rename and event attribute.
```
Comment (michalkarolik) [inline src/main/java/com/hltech/store/versioning/MappingBasedVersioning.java:16]: rename *an*

File: src/main/java/com/hltech/store/versioning/MultipleVersionsBasedVersioning.java [hunk 2]
```diff
@@ -3,7 +3,16 @@
...
 import java.util.Map;
 
-public class MultipleEventVersionPolicy<E> implements EventVersionPolicy<E> {
+/**
+ * In this strategy, multiple versions of the event has to bo supported in the application code.
+ * The application must contain knowledge of all deprecated event versions in order to support them.
+ * To avoid that consider using {@link UpcastingBasedVersioning}
+ *
+ * <p>Please note, that using this strategy is recommended only if you have one instance of you application running at the same time.
+ * Using this strategy in multi instance case, leads to the situation, where all instance must be updated
```
Comment (michalkarolik) [inline src/main/java/com/hltech/store/versioning/MultipleVersionsBasedVersioning.java:12]: Using this strategy in multi instance case leads to the situation, where all instances must be updated

**Review by michalkarolik — APPROVED** [review]

File: src/main/java/com/hltech/store/versioning/MultipleVersionsBasedVersioning.java [hunk 3]
```diff
@@ -1,21 +1,22 @@
-package com.hltech.store;
+package com.hltech.store.versioning;
 
-import java.util.Collection;
 import java.util.HashMap;
 import java.util.Map;
 
-public class SimpleEventTypeMapper<E> implements EventTypeMapper<E> {
+/**
+ * In this strategy, multiple versions of the event has to bo supported in the application code.
```
Comment (rydlu) [inline src/main/java/com/hltech/store/versioning/MultipleVersionsBasedVersioning.java:7]: has to be supported

File: src/main/java/com/hltech/store/versioning/MappingBasedVersioning.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+import java.util.Map;
+
+/**
+ * In this strategy, every event exists only in latest version, so that the application code has to support only one version of the event.
+ * The mapping strategy is based on three simple principles:
+ * - When attribute exists on both json and class then set the value from json
+ * - When attribute exists on json but not on class then do nothing
+ * - When attribute exists on class but not in json then set default value
+ *
+ * <p>This strategy is recommended when you have multiple instance of you application running at the same time,
```
Comment (rydlu) [inline src/main/java/com/hltech/store/versioning/MappingBasedVersioning.java:13]: when you have a multiple instance of your

File: src/main/java/com/hltech/store/versioning/UpcastingBasedVersioning.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
+package com.hltech.store.versioning;
+
+/**
+ * In this strategy, multiple versions of the event are up-casted to the latest version, using registered transformation.
+ * In opposite to {@link MultipleVersionsBasedVersioning} the application code has to support only the latest version of the event
+ *
+ * <p>Please note, that using this strategy is recommended only if you have one instance of you application running at the same time.
```
Comment (rydlu) [inline src/main/java/com/hltech/store/versioning/UpcastingBasedVersioning.java:7]: have one instance of **your** application

File: src/main/java/com/hltech/store/versioning/UpcastingBasedVersioning.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+ * Using this strategy in multi instance case, leads to the situation, where all instance must be updated
+ * to understand latest event version, before any instance produce it. For multi instance case consider using {@link MappingBasedVersioning}
```
Comment (rydlu) [inline src/main/java/com/hltech/store/versioning/UpcastingBasedVersioning.java:9]: before any instance **produces** it

File: src/main/java/com/hltech/store/versioning/MultipleVersionsBasedVersioning.java [hunk 7]
```diff
@@ -1,21 +1,22 @@
...
+ * The application must contain knowledge of all deprecated event versions in order to support them.
+ * To avoid that consider using {@link UpcastingBasedVersioning}
+ *
+ * <p>Please note, that using this strategy is recommended only if you have one instance of you application running at the same time.
```
Comment (rydlu) [inline src/main/java/com/hltech/store/versioning/MultipleVersionsBasedVersioning.java:11]: have one instance of **your** application

File: src/main/java/com/hltech/store/versioning/MultipleVersionsBasedVersioning.java [hunk 8]
```diff
@@ -1,21 +1,22 @@
...
+ * Using this strategy in multi instance case, leads to the situation, where all instance must be updated
+ * to understand latest event version, before any instance produce it. For multi instance case consider using {@link MappingBasedVersioning}
```
Comment (rydlu) [inline src/main/java/com/hltech/store/versioning/MultipleVersionsBasedVersioning.java:13]: before any instance **produces** it

File: src/main/java/com/hltech/store/versioning/WrappingBasedVersioning.java [hunk 9]
```diff
@@ -0,0 +1,30 @@
+package com.hltech.store.versioning;
+
+/**
+ * In this strategy, multiple versions of the event are wrapped, using registered wrapper.
+ * Wrapper is a class, that extends desirable event type, and provide all data directly from wrapped json.
```
Comment (rydlu) [inline src/main/java/com/hltech/store/versioning/WrappingBasedVersioning.java:5]: and **provides** all data

File: src/main/java/com/hltech/store/versioning/WrappingBasedVersioning.java [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+ *
+ * <p>This strategy is recommended when you have multiple instance of you application running at the same time,
```
Comment (rydlu) [inline src/main/java/com/hltech/store/versioning/WrappingBasedVersioning.java:7]: when you have **a** multiple instance of **your**

**Review by rydlu — APPROVED** [review]

**Pull request merged**